### DISCUSSION TITLE
How to tell to use format="date" instead of format="date-time"

### DISCUSSION THREAD

**Discussion by CathKilroy** [victools/jsonschema-generator#202] [Q&A] [answered]
Hi

I am using version 4.16.0.

In this version the standard format for Date objects, is "date-time". How can I tell the tool to us "date" instead?

Br Casper

**Comment by CarstenWickner** [comment, marked as answer]
HI @CathKilroy,

The default behavior you're describing comes from a combination of the following two [standard `Option`s](https://victools.github.io/jsonschema-generator/#generator-options) in your setup:
- `Option.ADDITIONAL_FIXED_TYPES`
- `Option.EXTRA_OPEN_API_FORMAT_VALUES`

The `"date-time"` value is defined here, since the `Date` class contains also a time portion:
https://github.com/victools/jsonschema-generator/blob/dc4662a7491485bd4a6ec08b0330e71d8573aefe/jsonschema-generator/src/main/java/com/github/victools/jsonschema/generator/impl/module/SimpleTypeModule.java#L85-L88

----

You have a few alternatives here. The simplest ones would be:
### 1. Keep your setup as-is and override the `"date-time"` with `"date"` with a `TypeAttributeOverride`
```java
configBuilder.forTypesInGeneral()
        .withTypeAttributeOverride(((collectedTypeAttributes, scope, context) -> {
            if (scope.getType().getErasedType() == Date.class) {
                collectedTypeAttributes.put(context.getKeyword(SchemaKeyword.TAG_FORMAT), "date");
            }
        }));
```
### 2. Apply an adjusted `SimpleTypeModule` instead of the one included by the standard `Option`s
```java
configBuilder.with(SimpleTypeModule.forPrimitiveAndAdditionalTypes().withStringType(Date.class, "date"));
```
### 3. Remove the `Option.EXTRA_OPEN_API_FORMAT_VALUES` and take full control over the `"format"` keyword
One would do that through a **[StringFormatResolver](https://victools.github.io/jsonschema-generator/#quot-format-quot-keyword)**:
```java
configBuilder.forTypesInGeneral()
    .withStringFormatResolver(scope -> scope.getType().getErasedType() == Date.class ? "date" : null);
// also need to take care of the other standard "format" values then
```
### 4. Override the whole subschema for the `Date` class with a "CustomDefinitionProvider" of your own:
```java
configBuilder.forTypesInGeneral()
        .withCustomDefinitionProvider((javaType, context) -> javaType.getErasedType() == Date.class
                ? new CustomDefinition(context.getGeneratorConfig().createObjectNode()
                        .put(context.getKeyword(SchemaKeyword.TAG_TYPE), context.getKeyword(SchemaKeyword.TAG_TYPE_STRING))
                        .put(context.getKeyword(SchemaKeyword.TAG_FORMAT), "date"),
                        CustomDefinition.DefinitionType.INLINE,
                        CustomDefinition.AttributeInclusion.YES)
                : null
                );
```

Due to the way the `Option.EXTRA_OPEN_API_FORMAT_VALUES` is being applied right now, a "StringFormatResolver" cannot override the `"format"` set in the `SimpleTypeModule` directly, as it's coming from the custom type definition directly. That could potentially be changed going forward, to also use a "StringFormatResolver" there, which could then be easily overridden.

At the moment number 1 and 2 appear to be the easiest ways to achieve what you want.
